src/Domain/DonationRepository.php

```diff
@@ -37,75 +45,50 @@ class DonationRepository extends SalesforceWriteProxyRepository
     private Matching\Adapter $matchingAdapter;
     /** @var Donation[] Tracks donations to persist outside the time-critical transaction / lock window */
     private array $queuedForPersist;
+
     public function setMatchingAdapter(Matching\Adapter $adapter): void
     {
         $this->matchingAdapter = $adapter;
     }
 
-    public function doCreate(SalesforceWriteProxy $proxy): bool
+    public function doCreate(AbstractStateChanged $changeMessage): bool
```

Comment:
why is a donation repository claiming to be able to handle creating an abstract state changed? Shouldn't it only handle the `DonationUpserted` ?

Reply:
I see it's because this is called by `\MatchBot\Domain\SalesforceWriteProxyRepository::push`. I think the fact that we're using one type here just to let the SA pass and then a narrower type in the assertion on the next line shows we're doing something a bit odd. It does seem like we're having quite a bit of extra work with the `SalesforceWriteProxyRepository` class that only has one child, but if we want it to keep working as it does without doing weird stuff then I think the way would be to remove the DonationUpserted class and replace it with `StateChanged<Donation>` . That allows \MatchBot\Domain\SalesforceWriteProxyRepository::push to have a param type of `StateChanged<T>`. Alternatively we could add a third template variable to SalesforceWriteProxyRepository, e.g. `@template S of AbstractStateChanged` and then the param type would just be `S`.

Reply:
There's also a much easier option - change the parameter types in `SalesforceWriteProxyRepository` from `AbstractStateChanged` to `DonationUpserted`. Of course that means those methods can't be used with anything other than a donation, but it doesn't currently need to be, and we could deal with that when we come to it.

Reply:
@bdsl IMO, the assertion is an OK trade-off for now until we decide (per the `AbstractStateChanged` class comment) whether we think the model is useful to expand mandates. My inclination would be to keep the workaround right now and then either use a more suitable type as you suggest if the pushing is similar enough that the abstraction is useful; or delete it if not. We'll be working on that reasonably soon I hope.